src/main/java/net/dries007/tfc/objects/te/TEQuern.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Work under Copyright. Licensed under the EUPL.
+ * See the project README.md and LICENSE.txt for more information.
+ */
+
+package net.dries007.tfc.objects.te;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.item.EntityXPOrb;
+import net.minecraft.entity.passive.EntityCow;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.EnumParticleTypes;
+import net.minecraft.util.ITickable;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.World;
+import net.minecraftforge.items.ItemStackHandler;
+
+import net.dries007.tfc.api.recipes.QuernRecipe;
+import net.dries007.tfc.objects.items.ItemHandstone;
+import net.dries007.tfc.objects.items.ItemsTFC;
+
+import static net.minecraft.init.SoundEvents.*;
+
+@ParametersAreNonnullByDefault
+public class TEQuern extends TEInventory implements ITickable
+{
+    public static final int SLOT_HANDSTONE = 0;
+    public static final int SLOT_INPUT = 1;
+    public static final int SLOT_OUTPUT = 2;
+
+    private int rotationTimer;
+    private boolean hasHandstone;
+
+    public TEQuern()
+    {
+        super(3);
+        rotationTimer = 0;
+    }
+
+    public ItemStack insertOrSwapItem(int slot, ItemStack playerStack)
+    {
+        ItemStack quernStack = inventory.getStackInSlot(slot);
+
+        if (quernStack.isEmpty() || (playerStack.isStackable() && quernStack.isStackable() && quernStack.getItem() == playerStack.getItem() && (!playerStack.getHasSubtypes() || playerStack.getMetadata() == quernStack.getMetadata()) && ItemStack.areItemStackTagsEqual(playerStack, quernStack)))
+        {
+            return inventory.insertItem(slot, playerStack, false);
+        }
+        inventory.setStackInSlot(slot, playerStack);
+        return quernStack;
+    }
+
+    public ItemStackHandler getInventory()
```

Comment:
Anything that calls this should instead call `tile.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null)`, and then null check the capability. This method should be removed